keps/sig-scheduling/20200612-simplified-topology-manager.md

```diff
@@ -0,0 +1,272 @@
+			return framework.NewStatus(framework.Unschedulable, fmt.Sprintf("Can't align container: %s", container.Name))
+		}
+	}
+	return nil
+}
+```
+## Accessing NodeResourceTopology CRD
+
+In order to allow the scheduler (deployed as a pod) to access NodeResourceTopology CRD instances, ClusterRole and ClusterRoleBinding would have to be configured as below:
+
+``` yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: noderesourcetopology-handler
+rules:
+- apiGroups: ["topology.node.k8s.io"]
+  resources: ["noderesourcetopologies"]
+  verbs: ["*"]
+- apiGroups: ["rbac.authorization.k8s.io"]
+  resources: ["*"]
+  verbs: ["*"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: handle-noderesourcetopology
+subjects:
+- kind: ServiceAccount
+  name: noderesourcetopology-account
+  namespace: default
+roleRef:
+  kind: ClusterRole
+  name: noderesourcetopology-handler
+  apiGroup: rbac.authorization.k8s.io
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: noderesourcetopology-account
+```
+
+`serviceAccountName: noderesourcetopology-account` would have to be added to the manifest file of the scheduler deployment file.
+
+# Use cases
+
+Numbers of kubernetes worker nodes on bare metal with NUMA topology. TopologyManager feature gate enabled on the nodes. In this configuration, the operator does not want that in the case of an unsatisfactory host topology, it should be re-scheduled for launch, but wants the scheduling to be successful the first time.
+
+# Known limitations
+
+Kube-scheduler makes an assumption about current resource usage on the worker node, since kube-scheduler knows which pod assigned to node. This assumption makes right after kube-scheduler choose a node. But in case of scheduling with NUMA topology only TopologyManager on the worker node knows exact NUMA node used by pod, this information about NUMA node delivers to kube-scheduler with latency. In this case kube-scheduler will not know actual NUMA topology until topology exporter will send it back. It could be mitigated if kube-scheduler in proposed plugin will add a hint on which NUMA id pod could be assigned, further Topology Manager on the worker node may take it into account.
+
+# Test plans
+
+It would be ensured that the components developed or modified for this feature can be easily tested.
+
+* Unit Tests
+
+Unit test for scheduler plugin (pkg/scheduler/framework/plugins/noderesources/topology_match.go)
+pkg/scheduler/framework/plugins/noderesources/topology_match_test.go which test the plugin.
+
+Separate tests for CRD informer also should be implemented.
+
+* Integration Tests and End-to-end tests
+
+Integration and End-to-end would Implementation of it does not constitute a difficulty, but requires appropriate multi-numa hardware for comprehensive testing of this feature. Comprehensive E2E testing of this would be done in order to graduate this feature from Alpha to Beta.
```

Comment:
this is the major blocker to knowing if the feature works.  this should be an alpha blocker.

Reply:
I think topology manager has it's own e2e tests, for example this job runs it:

Reply:
Can you be more specific on the testing procedures? If e2e is not appropriate, at least we need integration tests in the alpha phase to verify the whole workflow:

- default configuration (this plugin is disabled)
    - no side effect on basic scheduling flow (and performance)
    - no side effect no matter the CRD is installed or not
- enable this plugin
    - basic workflow of this feature works (decision by scheduler is admitted by kubelet)
    - basic negative path of this feature works (decision by scheduler is rejected by kubelet)
    - verify the behavior when the CRD is and isn't installed

Reply:
> Can you be more specific on the testing procedures? If e2e is not appropriate, at least we need integration tests in the alpha phase to verify the whole workflow:

@Huang-Wei  In relation to your comment here about having integration tests now and e2e tests in the next phase, what according to you is the main distinction between integration and e2e tests in this case as both are used for testing the end to end workflow of this feature. Can you help us understand the expectations in term of testing for a scheduling plugin being pushed as an alpha feature like the one we are proposing? Would appreciate if you help us gain some clarity on this.